test_util/dcos_api_session.py

```diff
@@ -30,11 +34,111 @@ def get_args_from_env():
+        self.credentials = credentials
+        self.auth_token = None
+        self.auth_cookie = None
+        self.auth = None
+
+    def authenticate(self, dcos_url: str):
+        logging.info('Attempting authentication')
+        # explicitly use a session with no user authentication for requesting auth headers
+        r = requests.post(path_join(dcos_url, '/acs/api/v1/auth/login'), json=self.credentials, auth=None)
+        r.raise_for_status()
+        logging.info('Received authentication blob: {}'.format(r.json()))
+        self.auth_token = r.json()['token']
+        self.auth_cookie = r.cookies['dcos-acs-auth-cookie']
+        logging.info('Authentication successful')
+        # Set requests auth
+        self.auth = DcosAuth(self.auth_token)
+
+
+class DcosAuth(requests.auth.AuthBase):
+    def __init__(self, auth_token: str):
+        self.auth_token = auth_token
+
+    def __call__(self, request):
+        request.headers['Authorization'] = 'token={}'.format(self.auth_token)
+        return request
+
+
+class ARNodeApiClientMixin:
+    def api_request(self, method, path_extension, scheme=None, host=None, query=None,
+                    fragment=None, port=None, node=None, **kwargs):
+        """ Communicating with a DC/OS cluster is done by default through Adminrouter.
+        Use this Mixin with an ApiClientSession that requires distinguishing between nodes.
+        Adminrouter has both a master and agent process and so this wrapper accepts a
+        node argument. node must be a host in self.master or self.all_slaves. If given,
+        the request will be made to the adminrouter endpoint for that node type
+        """
+        assert self.masters
+        assert self.all_slaves
+        if node is not None:
+            assert port is None, 'node is intended to retrieve port; cannot set both simultaneously'
+            assert host is None, 'node is intended to retrieve host; cannot set both simultaneously'
+            if node in self.masters:
+                # Nothing else to do, master adminrouter uses default HTTP (80) and HTTPS (443) ports
+                pass
+            elif node in self.all_slaves:
+                scheme = scheme if scheme is not None else self.default_url.scheme
+                if scheme == 'http':
+                    port = 61001
+                if scheme == 'https':
+                    port = 61002
+            else:
+                raise Exception('Node {} is not recognized within the DC/OS cluster'.format(node))
+            host = node
+        return super().api_request(method, path_extension, scheme=scheme, host=host,
+                                   query=query, fragment=fragment, port=port, **kwargs)
+
+
+class DcosApiSession(ARNodeApiClientMixin, ApiClientSession):
+    def __init__(self, dcos_url: str, masters: list, public_masters: list,
+                 slaves: list, public_slaves: list, default_os_user: str,
+                 auth_user=Optional[DcosUser]):
```

Comment:
Actually, None can be assigned to any object. 
This short snippet demonstrates the problem with this code. (foo is my recommendation, bar is the current approach).
```
$ cat a2.py
from typing import Optional

class DcosUser:
    def __init__(self, username):
        self.username = username
    def __str__(self):
        return self.username

class SomethingElse: pass

def foo(o:DcosUser = None):
    print(o)

foo()   # Valid
foo(None) # Valid
foo(DcosUser("s")) # Valid
foo(SomethingElse()) # Invalid


def bar(o=Optional[DcosUser]):
    print(o)

bar() # valid
bar(None) # valid. But is not using Optional, because it is setting to None
bar(DcosUser("s")) # valid
bar(SomethingElse()) # valid because it is setting to a different object. I don't think we want this.

$ mypy a2.py
a2.py:17: error: Argument 1 to "foo" has incompatible type "SomethingElse"; expected "DcosUser"
```

In addition, the original comment was pointing to this scenario. We use something like.

```
        if self.auth_user is not None:
```
In the way `self.auth` set, unless we specifically set the argument from the caller to `None` the the value is always not None.
```
>>> auth_user = Optional[object]
>>> auth_user is None
False
```